File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }

        int[] indegree = new int[numCourses];
        for (int[] prerequisite : prerequisites) {
            graph.get(prerequisite[1]).add(prerequisite[0]);
            indegree[prerequisite[0]]++;
        }

        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) queue.offer(i);
        }

        int[] result = new int[numCourses];
        int idx = 0;
        while (!queue.isEmpty()) {
            int cur = queue.poll();
            result[idx++] = cur;

            for (int next : graph.get(cur)) {
                indegree[next]--;
                if (indegree[next] == 0) queue.offer(next);
            }
        }

        if (idx != numCourses) return new int[0];

        return result;
    }
}

File: 0210-course-schedule-ii/0210-course-schedule-ii.kt
class Solution {
    fun findOrder(numCourses: Int, prerequisites: Array<IntArray>): IntArray {
        val graph = List(numCourses) { mutableListOf<Int>() }
        val indegree = IntArray(numCourses)
        for ((a, b) in prerequisites) {
            graph[b].add(a)
            indegree[a]++
        }

        val queue: Queue<Int> = LinkedList()
        for (i in 0 until numCourses) {
            if (indegree[i] == 0) queue.offer(i)
        }

        val result = IntArray(numCourses)
        var idx = 0
        while (queue.isNotEmpty()) {
            val cur = queue.poll()
            result[idx++] = cur

            for (next in graph[cur]) {
                indegree[next]--
                if (indegree[next] == 0) queue.offer(next)
            }
        }

        return if (idx == numCourses) result else intArrayOf()
    }
}